File: AbstractFactoryPattern/cpp/src/pizza.cpp
#include "header/pizza.h"

Pizza::Pizza(shared_ptr<vector<shared_ptr<Ingredient>>> ingredients){
    this->ingredients = ingredients;
}

CheesePizza::CheesePizza(shared_ptr<Cheese> cheese, shared_ptr<vector<shared_ptr<Ingredient>>> ingredients):Pizza(ingredients){
    this->cheese = cheese;
}

shared_ptr<Cheese> CheesePizza::getCheese(){
    return cheese;
}

void CheesePizza::setCheese(shared_ptr<Cheese> cheese){
    this->cheese = cheese;
}

string CheesePizza::showIngredients(){
    string ingredientsString = cheese->getInfo();
    for(auto ingredient: *ingredients){
        ingredientsString += (", " + ingredient->getInfo());
    }

    return ingredientsString;
}

CheesePizzaOfStoreA::CheesePizzaOfStoreA(shared_ptr<Cheese> cheese, shared_ptr<vector<shared_ptr<Ingredient>>> ingredients):CheesePizza(cheese, ingredients){
}

string CheesePizzaOfStoreA::getName(){
    return "store A cheese pizze";
}

CheesePizzaOfStoreB::CheesePizzaOfStoreB(shared_ptr<Cheese> cheese, shared_ptr<vector<shared_ptr<Ingredient>>> ingredients):CheesePizza(cheese, ingredients){
}

string CheesePizzaOfStoreB::getName(){
    return "store B cheese pizze";
}

File: AbstractFactoryPattern/python/PizzaStore/Pizza.py
from enum import Enum
import abc
from .Ingredient import *

class PizzaType(Enum):
    CHEESE_PIZZA = 0


class Pizza(metaclass=abc.ABCMeta):
    def __init__(self, ingredients:list[Ingredient]) -> None:
        self.ingredients = ingredients

    @abc.abstractmethod
    def getName(self) -> str:
        return NotImplemented

    @abc.abstractmethod
    def showIngredients(self) -> str:
        return NotImplemented
    

class CheesePizza(Pizza):
    def __init__(self, cheese:Cheese, ingredients:list[Ingredient]) -> None:
        super().__init__(ingredients)
        self.cheese = cheese


class CheesePizzeOfStoreA(CheesePizza):
    def __init__(self, cheese:Cheese, ingredients:list[Ingredient]) -> None:
        super().__init__(cheese, ingredients)
        
    def getName(self) -> str:
        return "store A cheese pizze"
    
    def showIngredients(self) -> str:
        ingredientsStr = self.cheese.getInfo()
        for ingredient in self.ingredients:
            ingredientsStr += (", " + ingredient.getInfo())

        return ingredientsStr


class CheesePizzeOfStoreB(CheesePizza):
    def __init__(self, cheese:Cheese, ingredients:list[Ingredient]) -> None:
        super().__init__(cheese, ingredients)

    def getName(self) -> str:
        return "store B cheese pizze"    

    def showIngredients(self) -> str:
        ingredientsStr = self.cheese.getInfo()
        for ingredient in self.ingredients:
            ingredientsStr += (", " + ingredient.getInfo())

        return ingredientsStr


